# Atelier NumPy — Capteurs IoT

**Contexte** : une entreprise a des capteurs (température, humidité, pression, consommation) dans plusieurs bâtiments. On prépare et analyse ces données avec NumPy avant de les envoyer à un futur système de Machine Learning.



## Partie 1 : Listes Python vs tableaux NumPy

Températures mesurées par un capteur : 24.5, 25.1, 26.3, 27.0, 25.8

1. Liste Python
2. Tableau NumPy équivalent
3. Multiplication par 2 sur les deux
4. Différence observée ?

In [1]:
import numpy as np

temperatures_liste = [24.5, 25.1, 26.3, 27.0, 25.8]
temperatures_liste

[24.5, 25.1, 26.3, 27.0, 25.8]

In [2]:
temperatures_array = np.array(temperatures_liste)
temperatures_array

array([24.5, 25.1, 26.3, 27. , 25.8])

In [3]:
temperatures_liste * 2

[24.5, 25.1, 26.3, 27.0, 25.8, 24.5, 25.1, 26.3, 27.0, 25.8]

In [4]:
temperatures_array * 2

array([49. , 50.2, 52.6, 54. , 51.6])

**Différence observée** :

- `liste * 2` **duplique** la liste (concatène la liste avec elle-même) → on obtient 10 éléments.
- `array * 2` applique la multiplication **élément par élément** (opération vectorisée) → on obtient toujours 5 éléments, chacun multiplié par 2.

C'est la différence fondamentale entre une liste Python (conteneur générique) et un tableau NumPy (structure numérique optimisée pour le calcul).

## Partie 2 : Création des données

1. Mesures de 10 capteurs, initialement à zéro
2. Tableau de 10 valeurs égales à 1
3. Tableau de 10 valeurs initialisées à 25
4. Matrice identité 5 × 5
5. Les 24 heures (0 à 23) d'une journée
6. 10 valeurs régulièrement réparties entre 20 °C et 30 °C
7. 100 mesures aléatoires, reproductibles

In [5]:
mesures_zero = np.zeros(10)
mesures_zero

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [6]:
mesures_un = np.ones(10)
mesures_un

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [7]:
mesures_25 = np.full(10, 25)
mesures_25

array([25, 25, 25, 25, 25, 25, 25, 25, 25, 25])

In [8]:
identite = np.eye(5,5)
identite

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [9]:
heures = np.arange(24)
heures

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23])

In [10]:
paliers = np.linspace(20, 30, 10)
paliers

array([20.        , 21.11111111, 22.22222222, 23.33333333, 24.44444444,
       25.55555556, 26.66666667, 27.77777778, 28.88888889, 30.        ])

Pour l'aléatoire **reproductible**, on utilise un générateur NumPy initialisé avec une
graine (*seed*) fixe : `np.random.default_rng(seed)`. Avec la même graine, on obtient
toujours la même séquence de nombres — utile pour que tout le monde obtienne exactement
les mêmes résultats en relançant le notebook.

In [11]:
rng = np.random.default_rng(42)
mesures_aleatoires = rng.normal(loc=25, scale=3, size=100)
mesures_aleatoires[:10]

array([25.91415124, 21.88004768, 27.25135359, 27.82169415, 19.14689443,
       21.09346148, 25.38352121, 24.05127222, 24.94959653, 22.44086822])

## Partie 3 : Exploration des tableaux

1. Créer aléatoirement un tableau représentant les 24 températures d'une journée
2. Déterminer : nombre de dimensions, nombre de valeurs, type des données, forme du tableau

On garde ce tableau `temperatures_jour` : il sera réutilisé dans les Parties 4 et 5.

In [12]:
temperatures_jour = rng.normal(loc=25, scale=6, size=24)
temperatures_jour

array([22.73102468, 32.79536979, 22.86241617, 29.42509341, 19.39829392,
       23.76737465, 19.29986767, 22.96580154, 30.04184882, 14.63607746,
       27.60654186, 26.42641361, 21.43510027, 16.32365287, 25.43277705,
       21.82304375, 26.39605727, 25.13111287, 34.61067335, 23.56386624,
       18.85901504, 26.07565381, 26.3199801 , 33.15512545])

In [13]:
temperatures_jour.ndim

1

In [14]:
temperatures_jour.size

24

In [15]:
temperatures_jour.dtype

dtype('float64')

In [16]:
temperatures_jour.shape

(24,)

- `ndim` : nombre d'axes/dimensions du tableau (ici 1, c'est un vecteur — on aura des matrices à `ndim=2` en Partie 6).
- `size` : nombre total de valeurs stockées (24, indépendamment du nombre de dimensions).
- `dtype` : type des éléments. NumPy stocke **un seul type homogène** par tableau (contrairement à une liste Python qui peut mélanger les types) — ici `float64`, un flottant sur 64 bits. C'est ce typage homogène qui permet à NumPy d'être rapide : il sait à l'avance combien d'octets occupe chaque élément.
- `shape` : un tuple donnant la taille de chaque dimension. Ici `(24,)` — un tuple à un seul élément car le tableau est 1D. En Partie 6, une matrice de 5 capteurs × 24 heures aura `shape = (5, 24)`.

## Partie 4 : Indexation et slicing

On réutilise `temperatures_jour` (24 valeurs, l'indice `i` correspond à l'heure `i`h,
de 0h à 23h). Afficher :

1. la première mesure
2. la dernière mesure
3. la température de 8h
4. les températures de 8h à 12h
5. les températures des 6 premières heures
6. les températures des heures paires

In [17]:
temperatures_jour[0]

np.float64(22.731024675763663)

In [18]:
temperatures_jour[-1]

np.float64(33.155125451442615)

In [19]:
temperatures_jour[8]

np.float64(30.041848824744374)

In [20]:
temperatures_jour[8:13]

array([30.04184882, 14.63607746, 27.60654186, 26.42641361, 21.43510027])

In [21]:
temperatures_jour[:6]

array([22.73102468, 32.79536979, 22.86241617, 29.42509341, 19.39829392,
       23.76737465])

In [22]:
temperatures_jour[::2]

array([22.73102468, 22.86241617, 19.39829392, 19.29986767, 30.04184882,
       27.60654186, 21.43510027, 25.43277705, 26.39605727, 34.61067335,
       18.85901504, 26.3199801 ])

Ce qu'il faut retenir :

- **Indexation simple** : `tableau[i]` accède à l'élément d'indice `i` — comme pour une liste
  Python, le premier élément est à l'indice `0` (donc `temperatures_jour[8]` correspond bien
  à 8h, pas à la 8ème mesure « humaine »).
- **Indice négatif** : `tableau[-1]` accède au dernier élément en partant de la fin, sans avoir
  à connaître la longueur du tableau. `-1` = dernier, `-2` = avant-dernier, etc.
- **Slicing `[début:fin]`** : renvoie une *sous-partie* du tableau, de l'indice `début` inclus
  à l'indice `fin` **exclu**. C'est pour ça qu'on écrit `[8:13]` (et pas `[8:12]`) pour obtenir
  les heures 8, 9, 10, 11 **et** 12.
- **Slicing avec pas `[début:fin:pas]`** : `[:6]` veut dire « du début jusqu'à l'indice 6 exclu »
  (donc les indices 0 à 5, soit les 6 premières heures). `[::2]` veut dire « du début à la fin,
  en avançant de 2 en 2 » — ici, comme les indices pairs correspondent aux heures paires, ça
  donne directement les températures de 0h, 2h, 4h, ..., 22h.
- Toutes ces opérations renvoient un **nouveau tableau** (sauf l'indexation simple qui renvoie
  un scalaire), et surtout : sur un tableau NumPy, le slicing renvoie une **vue** (pas une copie)
  qui partage la même mémoire que le tableau d'origine — modifier la vue modifierait aussi
  `temperatures_jour`. C'est une différence importante avec les listes Python, où trancher crée
  toujours une copie.

## Partie 5 : Filtrage booléen

Toujours sur `temperatures_jour`, on cherche à détecter des situations inhabituelles.

1. Températures supérieures à 30 °C
2. Températures « normales » (entre 20 °C et 30 °C)
3. `anomalies` : tableau de booléens, `True` si la température dépasse 35 °C
4. Nombre de températures anormales, à partir de `anomalies`

In [74]:
temperatures_jour[temperatures_jour > 30]

array([33.44489117, 30.49541525, 31.01654695, 33.02438864])

In [24]:
temperatures_jour[(temperatures_jour >= 20) & (temperatures_jour <= 30)]

array([22.73102468, 22.86241617, 29.42509341, 23.76737465, 22.96580154,
       27.60654186, 26.42641361, 21.43510027, 25.43277705, 21.82304375,
       26.39605727, 25.13111287, 23.56386624, 26.07565381, 26.3199801 ])

In [25]:
anomalies = temperatures_jour > 35
anomalies

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False])

In [26]:
anomalies.sum()

np.int64(0)

Le mécanisme central ici s'appelle le **masque booléen** (*boolean masking*) :

- `temperatures_jour > 30` ne renvoie **pas** un seul `True`/`False`, mais un tableau de
  24 booléens (un par élément), résultat de la comparaison élément par élément — encore
  la vectorisation vue en Partie 1.
- Utiliser ce tableau de booléens comme indice — `temperatures_jour[masque]` — ne garde
  que les éléments où le masque vaut `True`. C'est ce qu'on appelle le *filtrage booléen*.
- Pour combiner deux conditions, on utilise `&` (et) / `|` (ou) sur les tableaux, **pas**
  les mots-clés Python `and` / `or` qui ne fonctionnent que sur une seule valeur booléenne
  à la fois, pas sur un tableau entier. Et il faut **parenthéser chaque condition**
  (`(... >= 20) & (... <= 30)`), sinon Python évalue `&` avant `>=`/`<=` à cause des
  priorités des opérateurs, ce qui casse l'expression.
- `anomalies = temperatures_jour > 35` est lui-même un tableau de booléens de même taille
  que `temperatures_jour` — pas les valeurs anormales elles-mêmes, mais un « marqueur »
  `True`/`False` pour chaque heure.
- `anomalies.sum()` fonctionne parce qu'en Python, `True` vaut `1` et `False` vaut `0` :
  additionner un tableau de booléens revient donc à **compter les `True`**. Ici, le résultat
  est `0` : avec la graine `42` utilisée en Partie 3, aucune des 24 températures ne dépasse
  35 °C (la plus haute valeur observée était d'environ 33.4 °C) — c'est une journée sans
  anomalie, et c'est un résultat parfaitement valide.